#include "grupoestudodialog.h"
#include "ui_grupoestudodialog.h"
#include "chatmateria.h"
#include "perfildialog.h"
#include <QSqlQuery>
#include <QSqlError>
#include <QMessageBox>
#include <QDebug>
#include <QFrame>
#include <QLabel>
#include <QPushButton>
#include <QHBoxLayout>
#include <QVBoxLayout>
#include <QRandomGenerator>
#include <QEvent>
#include <QMouseEvent>

// ============================================================================
// CONSTRUTOR
// ============================================================================
GrupoEstudoDialog::GrupoEstudoDialog(QWidget *parent, const QString& username)
    : QDialog(parent)
    , ui(new Ui::GrupoEstudoDialog)
    , loggedInUsername(username)
    , telaAtual("materias")
{
    ui->setupUi(this);
    setWindowTitle("ğŸ“š Grupos de Estudo - EducaUTFPR");
    resize(1200, 800);

    setupDatabase();
    criarTabelasNecessarias();

    // Carrega as matÃ©rias no ComboBox de criar sala
    QSqlQuery query(dbConnection);
    query.exec("SELECT id_materia, nome FROM Materias ORDER BY nome");
    while (query.next()) {
        int id = query.value(0).toInt();
        QString nome = query.value(1).toString();
        ui->materiaComboBox->addItem(nome, id);
    }

    // Conecta o botÃ£o de confirmar criaÃ§Ã£o
    connect(ui->confirmarCriarButton, &QPushButton::clicked,
            this, &GrupoEstudoDialog::on_confirmarCriarButton_clicked);

    // Conecta o botÃ£o de entrar em grupo privado
    connect(ui->entrarGrupoButton, &QPushButton::clicked,
            this, &GrupoEstudoDialog::onEntrarGrupoPrivado);

    // Conecta os botÃµes de navegaÃ§Ã£o
    connect(ui->materiasButton, &QPushButton::clicked,
            this, &GrupoEstudoDialog::on_materiasButton_clicked);
    connect(ui->gruposButton, &QPushButton::clicked,
            this, &GrupoEstudoDialog::on_gruposButton_clicked);
    connect(ui->criarButton, &QPushButton::clicked,
            this, &GrupoEstudoDialog::on_criarButton_clicked);

    // Filtro de categoria
    connect(ui->categoriaComboBox, QOverload<int>::of(&QComboBox::currentIndexChanged),
            this, &GrupoEstudoDialog::on_categoriaComboBox_currentIndexChanged);

    // Configura os eventos de clique na barra de navegaÃ§Ã£o
    setupNavigationBar();

    // Mostra a tela inicial (MatÃ©rias)
    mostrarTela("materias");
}

// ============================================================================
// DESTRUTOR
// ============================================================================
GrupoEstudoDialog::~GrupoEstudoDialog()
{
    delete ui;
}

// ============================================================================
// SETUP DATABASE
// ============================================================================
void GrupoEstudoDialog::setupDatabase()
{
    dbConnection = QSqlDatabase::database("qt_sql_default_connection");

    if (!dbConnection.isOpen()) {
        qDebug() << "[GrupoEstudoDialog] ERRO: Banco de dados nÃ£o estÃ¡ aberto.";
    }
}

// ============================================================================
// CRIAR TABELAS NECESSÃRIAS
// ============================================================================
void GrupoEstudoDialog::criarTabelasNecessarias()
{
    QSqlQuery query(dbConnection);

    // Tabela de MatÃ©rias
    query.exec(
        "CREATE TABLE IF NOT EXISTS Materias ("
        "id_materia INTEGER PRIMARY KEY AUTOINCREMENT, "
        "nome TEXT NOT NULL, "
        "categoria TEXT NOT NULL, "
        "icone TEXT, "
        "cor TEXT)"
        );

    // Tabela de Salas de Estudo
    query.exec(
        "CREATE TABLE IF NOT EXISTS Salas_Estudo ("
        "id_sala INTEGER PRIMARY KEY AUTOINCREMENT, "
        "id_materia INTEGER NOT NULL, "
        "nome_sala TEXT NOT NULL, "
        "codigo_sala TEXT UNIQUE NOT NULL, "
        "tipo TEXT NOT NULL, "  // 'publica' ou 'privada'
        "senha TEXT, "
        "max_participantes INTEGER DEFAULT 10, "
        "data_criacao DATETIME DEFAULT CURRENT_TIMESTAMP, "
        "FOREIGN KEY (id_materia) REFERENCES Materias(id_materia))"
        );

    // Tabela de Participantes
    query.exec(
        "CREATE TABLE IF NOT EXISTS Participantes_Sala ("
        "id_participante INTEGER PRIMARY KEY AUTOINCREMENT, "
        "id_sala INTEGER NOT NULL, "
        "id_usuario INTEGER NOT NULL, "
        "data_entrada DATETIME DEFAULT CURRENT_TIMESTAMP, "
        "FOREIGN KEY (id_sala) REFERENCES Salas_Estudo(id_sala), "
        "FOREIGN KEY (id_usuario) REFERENCES Usuario(id_usuario), "
        "UNIQUE(id_sala, id_usuario))"
        );

    // Tabela de Mensagens do Chat
    query.exec(
        "CREATE TABLE IF NOT EXISTS Mensagens_Chat ("
        "id_mensagem INTEGER PRIMARY KEY AUTOINCREMENT, "
        "id_sala INTEGER NOT NULL, "
        "id_usuario INTEGER NOT NULL, "
        "mensagem TEXT NOT NULL, "
        "data_envio DATETIME DEFAULT CURRENT_TIMESTAMP, "
        "FOREIGN KEY (id_sala) REFERENCES Salas_Estudo(id_sala), "
        "FOREIGN KEY (id_usuario) REFERENCES Usuario(id_usuario))"
        );

    // Popula matÃ©rias de exemplo (se nÃ£o existirem)
    query.exec("SELECT COUNT(*) FROM Materias");
    if (query.next() && query.value(0).toInt() == 0) {
        popularMateriasCompletas();
    }
}

// ============================================================================
// POPULAR MATÃ‰RIAS COMPLETAS
// ============================================================================
void GrupoEstudoDialog::popularMateriasCompletas()
{
    QSqlQuery query(dbConnection);

    QList<QStringList> materias = {
        // MATEMÃTICA
        {"CÃ¡lculo Diferencial e Integral 1", "MatemÃ¡tica", "ğŸ“", "#FF6B6B"},
        {"CÃ¡lculo Diferencial e Integral 2", "MatemÃ¡tica", "ğŸ“", "#FF8787"},
        {"CÃ¡lculo Diferencial e Integral 3", "MatemÃ¡tica", "ğŸ“", "#FFA07A"},
        {"Ãlgebra Linear", "MatemÃ¡tica", "ğŸ“Š", "#FFB6B9"},
        {"Geometria AnalÃ­tica", "MatemÃ¡tica", "ğŸ“", "#FFCCCC"},
        {"MatemÃ¡tica Discreta", "MatemÃ¡tica", "ğŸ”¢", "#FF9999"},
        {"EquaÃ§Ãµes Diferenciais OrdinÃ¡rias", "MatemÃ¡tica", "ğŸ“ˆ", "#FFB347"},
        {"CÃ¡lculo NumÃ©rico", "MatemÃ¡tica", "ğŸ”¢", "#FFAB73"},
        {"Probabilidade e EstatÃ­stica", "MatemÃ¡tica", "ğŸ“Š", "#FFC9A0"},
        {"AnÃ¡lise de Sistemas Lineares", "MatemÃ¡tica", "ğŸ“ˆ", "#FFD9B3"},

        // FÃSICA
        {"FÃ­sica TeÃ³rica 1", "FÃ­sica", "âš›ï¸", "#4ECDC4"},
        {"FÃ­sica TeÃ³rica 2", "FÃ­sica", "âš›ï¸", "#45B7D1"},
        {"FÃ­sica TeÃ³rica 3", "FÃ­sica", "âš›ï¸", "#5AD5E5"},
        {"FÃ­sica Experimental 1", "FÃ­sica", "ğŸ§ª", "#70E0F0"},
        {"FÃ­sica Experimental 2", "FÃ­sica", "ğŸ§ª", "#85EBFA"},

        // PROGRAMAÃ‡ÃƒO
        {"Fundamentos de ProgramaÃ§Ã£o 1", "ProgramaÃ§Ã£o", "ğŸ’»", "#F38181"},
        {"Fundamentos de ProgramaÃ§Ã£o 2", "ProgramaÃ§Ã£o", "ğŸ’»", "#FF8FA3"},
        {"ProgramaÃ§Ã£o Orientada a Objetos", "ProgramaÃ§Ã£o", "ğŸ²", "#FFADAD"},
        {"Estrutura de Dados 1", "ProgramaÃ§Ã£o", "ğŸ“¦", "#FFD6A5"},
        {"Estrutura de Dados 2", "ProgramaÃ§Ã£o", "ğŸ“¦", "#FDFFB6"},
        {"Compiladores", "ProgramaÃ§Ã£o", "ğŸ“", "#CAFFBF"},
        {"Desenvolvimento de AplicaÃ§Ãµes Web", "ProgramaÃ§Ã£o", "ğŸŒ", "#9BF6FF"},

        // BANCO DE DADOS
        {"Banco de Dados", "Banco de Dados", "ğŸ—„ï¸", "#A0C4FF"},

        // REDES E SISTEMAS
        {"Sistemas Operacionais", "Redes e Sistemas", "ğŸ’¾", "#BDB2FF"},
        {"Redes de Computadores", "Redes e Sistemas", "ğŸŒ", "#FFC6FF"},
        {"ComunicaÃ§Ã£o de Dados", "Redes e Sistemas", "ğŸ“¡", "#E7C6FF"},
        {"Sistemas DistribuÃ­dos", "Redes e Sistemas", "â˜ï¸", "#C8B6FF"},
        {"SeguranÃ§a e Auditoria de Sistemas", "Redes e Sistemas", "ğŸ”", "#D4A5FF"},

        // ENGENHARIA
        {"IntroduÃ§Ã£o Ã  Engenharia de ComputaÃ§Ã£o", "Engenharia", "ğŸ’¡", "#FFE66D"},
        {"Arquitetura e OrganizaÃ§Ã£o de Computadores", "Engenharia", "ğŸ–¥ï¸", "#FFEB99"},
        {"Circuitos Digitais", "Engenharia", "ğŸ“Œ", "#FFF4B8"},
        {"Sistemas Digitais", "Engenharia", "ğŸ”§", "#FFFBD4"},
        {"EletrÃ´nica A", "Engenharia", "ğŸ“Œ", "#FDE4CF"},
        {"EletrÃ´nica B", "Engenharia", "ğŸ“Œ", "#FFCFD2"},
        {"AnÃ¡lise de Circuitos ElÃ©tricos 1", "Engenharia", "âš¡", "#F1C0E8"},
        {"Materiais e Equipamentos ElÃ©tricos", "Engenharia", "âš¡", "#CFBAF0"},
        {"Desenho TÃ©cnico", "Engenharia", "âœï¸", "#A3C4F3"},
        {"Fundamentos de Controle", "Engenharia", "ğŸ›ï¸", "#90DBF4"},
        {"Controle Digital", "Engenharia", "ğŸ®", "#8EECF5"},
        {"LÃ³gica ReconfigurÃ¡vel", "Engenharia", "ğŸ”§", "#98F5E1"},
        {"Sistemas Microcontrolados", "Engenharia", "âš™ï¸", "#B9FBC0"},
        {"Sistemas Embarcados", "Engenharia", "ğŸ”§", "#C7CEEA"},
        {"InstrumentaÃ§Ã£o EletrÃ´nica", "Engenharia", "ğŸ“¡", "#FFDAB9"},
        {"Processamento Digital de Sinais", "Engenharia", "ğŸ“Š", "#FFDFBA"},
        {"Oficina de IntegraÃ§Ã£o 1", "Engenharia", "ğŸ› ï¸", "#FFFFBA"},
        {"Oficina de IntegraÃ§Ã£o 2", "Engenharia", "ğŸ› ï¸", "#BAFFC9"},

        // QUÃMICA
        {"QuÃ­mica Geral", "QuÃ­mica", "âš—ï¸", "#95E1D3"},
        {"QuÃ­mica Experimental", "QuÃ­mica", "ğŸ§ª", "#A8E6CF"},

        // COMPUTAÃ‡ÃƒO TEÃ“RICA
        {"Teoria da ComputaÃ§Ã£o", "ProgramaÃ§Ã£o", "ğŸ§®", "#DCEDC1"},
        {"IntroduÃ§Ã£o Ã  LÃ³gica para ComputaÃ§Ã£o", "ProgramaÃ§Ã£o", "ğŸ§ ", "#FFD3B6"},
        {"Sistemas Inteligentes 1", "ProgramaÃ§Ã£o", "ğŸ¤–", "#FFAAA5"},

        // ENGENHARIA DE SOFTWARE
        {"Engenharia de Software", "ProgramaÃ§Ã£o", "ğŸ—‚ï¸", "#FF8B94"},

        // HUMANAS E SOCIAIS
        {"ComunicaÃ§Ã£o LinguÃ­stica", "Humanas e Sociais", "ğŸ“", "#A8DADC"},
        {"InglÃªs Instrumental", "Humanas e Sociais", "ğŸŒ", "#457B9D"},
        {"Metodologia de Pesquisa", "Humanas e Sociais", "ğŸ“š", "#1D3557"},
        {"CiÃªncias do Ambiente", "Humanas e Sociais", "ğŸŒ±", "#2A9D8F"},
        {"Economia", "Humanas e Sociais", "ğŸ’°", "#E76F51"},
        {"Empreendedorismo", "Humanas e Sociais", "ğŸ’¡", "#F4A261"},
        {"RelaÃ§Ãµes Humanas e LideranÃ§a", "Humanas e Sociais", "ğŸ¤", "#E9C46A"},
        {"Meio Ambiente e Sociedade", "Humanas e Sociais", "ğŸŒ", "#2A9134"},
        {"Qualidade de Vida", "Humanas e Sociais", "ğŸ’š", "#52B788"},
        {"AptidÃ£o FÃ­sica", "Humanas e Sociais", "ğŸƒ", "#74C69D"},
        {"Libras 1", "Humanas e Sociais", "ğŸ‘‹", "#95D5B2"},

        // TCC E ESTÃGIO
        {"Trabalho de ConclusÃ£o de Curso 1", "Engenharia", "ğŸ“„", "#B7E4C7"},
        {"Trabalho de ConclusÃ£o de Curso 2", "Engenharia", "ğŸ“„", "#D8F3DC"},
        {"EstÃ¡gio Curricular ObrigatÃ³rio", "Engenharia", "ğŸ’¼", "#E8F5E9"},

        // ATIVIDADES COMPLEMENTARES
        {"Atividades Complementares", "Humanas e Sociais", "ğŸ¯", "#C7CEEA"}
    };

    for (const auto& mat : materias) {
        query.prepare("INSERT INTO Materias (nome, categoria, icone, cor) VALUES (?, ?, ?, ?)");
        query.addBindValue(mat[0]);
        query.addBindValue(mat[1]);
        query.addBindValue(mat[2]);
        query.addBindValue(mat[3]);
        query.exec();
    }

    qDebug() << "âœ… MatÃ©rias populadas com sucesso!";
}

// ============================================================================
// CRIAR SALA (AÃ‡ÃƒO DO BOTÃƒO CONFIRMAR)
// ============================================================================
void GrupoEstudoDialog::on_confirmarCriarButton_clicked()
{
    if (!validarCriacaoSala()) {
        return;
    }

    QString nomeSala = ui->nomeSalaEdit->text().trimmed();
    int idMateria = ui->materiaComboBox->currentData().toInt();
    QString tipo = ui->tipoPublicaRadio->isChecked() ? "publica" : "privada";
    QString senha = ui->senhaEdit->text();
    int maxParticipantes = ui->maxParticipantesSpinBox->value();
    QString codigo = gerarCodigoSala();

    QSqlQuery query(dbConnection);
    query.prepare(
        "INSERT INTO Salas_Estudo (id_materia, nome_sala, codigo_sala, tipo, senha, max_participantes) "
        "VALUES (?, ?, ?, ?, ?, ?)"
        );

    // Binds das variÃ¡veis coletadas acima
    query.addBindValue(idMateria);
    query.addBindValue(nomeSala);
    query.addBindValue(codigo);
    query.addBindValue(tipo);
    query.addBindValue(tipo == "privada" ? senha : QVariant());
    query.addBindValue(maxParticipantes);

    if (query.exec()) {
        int idSala = query.lastInsertId().toInt();

        // Adiciona criador como participante
        int idUsuario = getIdUsuario(loggedInUsername);

        QSqlQuery pQuery(dbConnection);
        pQuery.prepare("INSERT INTO Participantes_Sala (id_sala, id_usuario) VALUES (?, ?)");
        pQuery.addBindValue(idSala);
        pQuery.addBindValue(idUsuario);
        pQuery.exec();

        QMessageBox::information(this, "âœ… Grupo Criado!",
                                 QString("CÃ³digo do grupo: %1\n%2")
                                     .arg(codigo)
                                     .arg(tipo == "privada" ? "Compartilhe com cuidado!" : ""));

        // Limpa os campos
        ui->nomeSalaEdit->clear();
        ui->senhaEdit->clear();

        // Volta para a tela de grupos
        mostrarTela("grupos");
    } else {
        QMessageBox::critical(this, "âŒ Erro",
                              "Erro ao criar grupo: " + query.lastError().text());
    }
}

// ============================================================================
// AUXILIARES E NAVEGAÃ‡ÃƒO
// ============================================================================
int GrupoEstudoDialog::getIdUsuario(const QString& username)
{
    QSqlQuery query(dbConnection);
    query.prepare("SELECT id_usuario FROM Usuario WHERE usuario = ?");
    query.addBindValue(username);

    if (query.exec() && query.next()) {
        return query.value(0).toInt();
    }
    return -1;
}

void GrupoEstudoDialog::mostrarTela(const QString& tela)
{
    telaAtual = tela;

    if (tela == "materias") {
        ui->stackedWidget->setCurrentIndex(0);
        carregarMateriasDaCategoria(ui->categoriaComboBox->currentText());
    } else if (tela == "grupos") {
        ui->stackedWidget->setCurrentIndex(1);
        carregarSalasPublicas();
    } else if (tela == "criar") {
        ui->stackedWidget->setCurrentIndex(2);
    }
}

void GrupoEstudoDialog::carregarMateriasDaCategoria(const QString& categoria)
{
    QWidget *container = ui->materiasContainer;
    QVBoxLayout *layout = qobject_cast<QVBoxLayout*>(container->layout());

    if (!layout) {
        layout = new QVBoxLayout(container);
        layout->setSpacing(12);
        layout->setContentsMargins(15, 15, 15, 15);
        container->setLayout(layout);
    }

    // Limpa o layout
    QLayoutItem *item;
    while ((item = layout->takeAt(0)) != nullptr) {
        delete item->widget();
        delete item;
    }

    // Busca matÃ©rias
    QSqlQuery query(dbConnection);
    QString queryStr = "SELECT id_materia, nome, icone, cor FROM Materias";

    if (!categoria.contains("Todas")) {
        // Remove o emoji do inÃ­cio se houver (lÃ³gica simples)
        QString categoriaLimpa = categoria;
        int primeiroEspaco = categoriaLimpa.indexOf(' ');
        if (primeiroEspaco != -1) {
            categoriaLimpa = categoriaLimpa.mid(primeiroEspaco + 1).trimmed();
        }

        queryStr += " WHERE categoria = ?";
        query.prepare(queryStr);
        query.addBindValue(categoriaLimpa);
    } else {
        query.prepare(queryStr);
    }

    query.exec();

    while (query.next()) {
        int id = query.value(0).toInt();
        QString nome = query.value(1).toString();
        QString icone = query.value(2).toString();
        QString cor = query.value(3).toString();

        QPushButton *btn = criarBotaoMateria(id, nome, icone, cor);
        layout->addWidget(btn);
    }

    layout->addStretch();
}

QPushButton* GrupoEstudoDialog::criarBotaoMateria(int idMateria, const QString& nome,
                                                  const QString& icone, const QString& cor)
{
    QPushButton *btn = new QPushButton();
    btn->setText(QString("%1  %2").arg(icone).arg(nome));
    btn->setMinimumHeight(70);
    btn->setCursor(Qt::PointingHandCursor);

    QString corBase = cor;
    QString corHover = corBase;

    // Efeito simples de hover escurecendo levemente
    if (corHover.startsWith("#FF")) {
        corHover.replace(0, 3, "#DD");
    } else if (corHover.contains("FF")) {
        corHover.replace("FF", "DD");
    }

    btn->setStyleSheet(
        QString("QPushButton {"
                "   background-color: %1;"
                "   color: white;"
                "   border: none;"
                "   border-radius: 12px;"
                "   padding: 15px 25px;"
                "   font-size: 18px;"
                "   font-weight: bold;"
                "   text-align: left;"
                "}"
                "QPushButton:hover {"
                "   background-color: %2;"
                "   transform: scale(1.02);"
                "}").arg(corBase).arg(corHover)
        );

    connect(btn, &QPushButton::clicked, [this, idMateria]() {
        onMateriaClicked(idMateria);
    });

    return btn;
}

void GrupoEstudoDialog::onMateriaClicked(int idMateria)
{
    // Busca o nome da matÃ©ria
    QSqlQuery query(dbConnection);
    query.prepare("SELECT nome FROM Materias WHERE id_materia = ?");
    query.addBindValue(idMateria);

    if (!query.exec() || !query.next()) {
        return;
    }

    QString nomeMateria = query.value(0).toString();

    // Busca a sala geral da matÃ©ria (ou cria se nÃ£o existir)
    query.prepare("SELECT id_sala, codigo_sala, nome_sala FROM Salas_Estudo WHERE id_materia = ? AND tipo = 'publica'");
    query.addBindValue(idMateria);

    int idSala = 0;
    QString nomeSala = "";

    if (query.exec() && query.next()) {
        idSala = query.value(0).toInt();
        nomeSala = query.value(2).toString();
    } else {
        // Cria sala geral automaticamente
        QString codigo = gerarCodigoSala();
        query.prepare(
            "INSERT INTO Salas_Estudo (id_materia, nome_sala, codigo_sala, tipo, max_participantes) "
            "VALUES (?, ?, ?, 'publica', 50)"
            );
        query.addBindValue(idMateria);
        query.addBindValue("Chat Geral - " + nomeMateria);
        query.addBindValue(codigo);
        query.exec();

        idSala = query.lastInsertId().toInt();
        nomeSala = "Chat Geral - " + nomeMateria;
    }

    // Adiciona usuÃ¡rio como participante
    int idUsuario = getIdUsuario(loggedInUsername);
    query.prepare("INSERT OR IGNORE INTO Participantes_Sala (id_sala, id_usuario) VALUES (?, ?)");
    query.addBindValue(idSala);
    query.addBindValue(idUsuario);
    query.exec();

    // Abre o chat
    ChatMateria *chat = new ChatMateria(this, loggedInUsername, idSala, nomeSala);
    chat->exec();
    delete chat;
}

void GrupoEstudoDialog::carregarSalasPublicas()
{
    QWidget *container = ui->salasContainer;
    QVBoxLayout *layout = qobject_cast<QVBoxLayout*>(container->layout());

    if (!layout) {
        layout = new QVBoxLayout(container);
        layout->setSpacing(12);
        layout->setContentsMargins(15, 15, 15, 15);
        container->setLayout(layout);
    }

    // Limpa
    QLayoutItem *item;
    while ((item = layout->takeAt(0)) != nullptr) {
        delete item->widget();
        delete item;
    }

    // Busca salas pÃºblicas
    QSqlQuery query(dbConnection);
    query.exec(
        "SELECT s.id_sala, s.codigo_sala, s.nome_sala, s.tipo, s.max_participantes, "
        "COUNT(p.id_usuario) as num_participantes "
        "FROM Salas_Estudo s "
        "LEFT JOIN Participantes_Sala p ON s.id_sala = p.id_sala "
        "WHERE s.tipo = 'publica' "
        "GROUP BY s.id_sala "
        "ORDER BY num_participantes DESC"
        );

    while (query.next()) {
        int idSala = query.value(0).toInt();
        QString codigoSala = query.value(1).toString();
        QString nomeSala = query.value(2).toString();
        QString tipo = query.value(3).toString();
        int maxParticipantes = query.value(4).toInt();
        int numParticipantes = query.value(5).toInt();

        QFrame *card = criarCardSala(idSala, codigoSala, nomeSala, tipo,
                                     numParticipantes, maxParticipantes);
        layout->addWidget(card);
    }

    layout->addStretch();
}

QFrame* GrupoEstudoDialog::criarCardSala(int idSala, const QString& codigoSala,
                                         const QString& nomeSala, const QString& tipo,
                                         int numParticipantes, int maxParticipantes)
{
    QFrame *card = new QFrame();
    card->setFrameShape(QFrame::StyledPanel);
    card->setMinimumHeight(100);
    card->setStyleSheet(
        "QFrame {"
        "   background-color: #423738;"
        "   border-left: 5px solid #F4B315;"
        "   border-radius: 10px;"
        "   padding: 15px;"
        "   margin: 5px;"
        "}"
        "QFrame:hover {"
        "   background-color: #524447;"
        "}"
        );

    QHBoxLayout *mainLayout = new QHBoxLayout(card);

    // InformaÃ§Ãµes
    QVBoxLayout *infoLayout = new QVBoxLayout();

    QLabel *nomeLabel = new QLabel(nomeSala);
    nomeLabel->setStyleSheet("font-size: 16px; font-weight: bold; color: #F4B315;");

    QLabel *infoLabel = new QLabel(
        QString("ğŸ‘¥ %1/%2 participantes | ğŸ”‘ %3")
            .arg(numParticipantes)
            .arg(maxParticipantes)
            .arg(codigoSala)
        );
    infoLabel->setStyleSheet("font-size: 12px; color: #D3AF35;");

    infoLayout->addWidget(nomeLabel);
    infoLayout->addWidget(infoLabel);

    // BotÃ£o Entrar
    QPushButton *entrarBtn = new QPushButton("ğŸ”¥ Entrar");
    entrarBtn->setMinimumWidth(100);
    entrarBtn->setStyleSheet(
        "QPushButton {"
        "   background-color: #F4B315;"
        "   color: #1A161A;"
        "   border: none;"
        "   border-radius: 8px;"
        "   padding: 10px 20px;"
        "   font-weight: bold;"
        "}"
        "QPushButton:hover { background-color: #D3AF35; }"
        );

    connect(entrarBtn, &QPushButton::clicked, [this, idSala]() {
        onSalaClicked(idSala);
    });

    mainLayout->addLayout(infoLayout, 1);
    mainLayout->addWidget(entrarBtn);

    return card;
}

void GrupoEstudoDialog::onSalaClicked(int idSala)
{
    // Busca informaÃ§Ãµes da sala
    QSqlQuery query(dbConnection);
    query.prepare("SELECT nome_sala FROM Salas_Estudo WHERE id_sala = ?");
    query.addBindValue(idSala);

    if (!query.exec() || !query.next()) {
        return;
    }

    QString nomeSala = query.value(0).toString();

    // Adiciona usuÃ¡rio como participante
    int idUsuario = getIdUsuario(loggedInUsername);
    query.prepare("INSERT OR IGNORE INTO Participantes_Sala (id_sala, id_usuario) VALUES (?, ?)");
    query.addBindValue(idSala);
    query.addBindValue(idUsuario);
    query.exec();

    // Abre o chat
    ChatMateria *chat = new ChatMateria(this, loggedInUsername, idSala, nomeSala);
    chat->exec();
    delete chat;
}

// ============================================================================
// NAVEGAÃ‡ÃƒO E EVENTOS
// ============================================================================
void GrupoEstudoDialog::on_materiasButton_clicked()
{
    mostrarTela("materias");
}

void GrupoEstudoDialog::on_gruposButton_clicked()
{
    mostrarTela("grupos");
}

void GrupoEstudoDialog::on_criarButton_clicked()
{
    mostrarTela("criar");
}

void GrupoEstudoDialog::on_categoriaComboBox_currentIndexChanged(int index)
{
    Q_UNUSED(index);
    carregarMateriasDaCategoria(ui->categoriaComboBox->currentText());
}

void GrupoEstudoDialog::setupNavigationBar()
{
    ui->homeButton->installEventFilter(this);
    ui->perfilButton->installEventFilter(this);
}

bool GrupoEstudoDialog::eventFilter(QObject *obj, QEvent *event)
{
    if (obj == ui->homeButton && event->type() == QEvent::MouseButtonPress) {
        this->close();
        return true;
    }

    if (obj == ui->perfilButton && event->type() == QEvent::MouseButtonPress) {
        PerfilDialog *perfil = new PerfilDialog(this, loggedInUsername);
        perfil->exec();
        delete perfil;
        return true;
    }

    return QDialog::eventFilter(obj, event);
}

void GrupoEstudoDialog::onEntrarGrupoPrivado()
{
    QString codigo = ui->codigoGrupoEdit->text().trimmed();
    QString senha = ui->senhaGrupoEdit->text();

    if (codigo.isEmpty()) {
        QMessageBox::warning(this, "âš ï¸ Campo Vazio",
                             "Por favor, digite o cÃ³digo do grupo!");
        return;
    }

    QSqlQuery query(dbConnection);
    query.prepare(
        "SELECT id_sala, tipo, senha, nome_sala FROM Salas_Estudo "
        "WHERE codigo_sala = ?"
        );
    query.addBindValue(codigo);

    if (!query.exec() || !query.next()) {
        QMessageBox::warning(this, "âš ï¸ Grupo NÃ£o Encontrado",
                             "NÃ£o existe nenhum grupo com este cÃ³digo!");
        return;
    }

    int idSala = query.value(0).toInt();
    QString tipo = query.value(1).toString();
    QString senhaCorreta = query.value(2).toString();
    QString nomeSala = query.value(3).toString();

    if (tipo == "privada" && senha != senhaCorreta) {
        QMessageBox::warning(this, "âš ï¸ Senha Incorreta",
                             "A senha digitada estÃ¡ incorreta!");
        return;
    }

    int idUsuario = getIdUsuario(loggedInUsername);
    query.prepare("INSERT OR IGNORE INTO Participantes_Sala (id_sala, id_usuario) VALUES (?, ?)");
    query.addBindValue(idSala);
    query.addBindValue(idUsuario);
    query.exec();

    ui->codigoGrupoEdit->clear();
    ui->senhaGrupoEdit->clear();

    ChatMateria *chat = new ChatMateria(this, loggedInUsername, idSala, nomeSala);
    chat->exec();
    delete chat;
}

QString GrupoEstudoDialog::gerarCodigoSala()
{
    QString codigo;
    for (int i = 0; i < 5; i++) {
        codigo += QString::number(QRandomGenerator::global()->bounded(10));
    }
    return codigo;
}

bool GrupoEstudoDialog::validarCriacaoSala()
{
    if (ui->nomeSalaEdit->text().trimmed().isEmpty()) {
        QMessageBox::warning(this, "âš ï¸ Campo Vazio", "Digite o nome do grupo!");
        return false;
    }

    if (ui->tipoPrivadaRadio->isChecked() && ui->senhaEdit->text().isEmpty()) {
        QMessageBox::warning(this, "âš ï¸ Senha ObrigatÃ³ria",
                             "Grupos privados precisam de senha!");
        return false;
    }

    return true;
}
